README
Debian packaging for The Librarian File System (LFS) Suite

The "upstream" branch of this repo contains the source code and detailed
documentation.  This "debian" branch is to be combined with "upstream" to
enable the creation of Debian packages.  In particular, a "debian/gbp.conf"
exists to control the execution of gbp-buildpackage.  Per "man gbp.conf"
it should override any system-wide or personal settings.

To make the binary packages python3-tm-librarian, tm-librarian, and tm-lfs:

1. Check for ~/.devscripts as it may contain conflicting settings, particularly
   the DEBUILD_DPKG_BUILDPACKAGE_OPTS variable.  "man debuild" has more info.
   Rename or remove the file if you are unsure of its interaction.

2. Start from the "debian" branch and insure there is no "master" branch:

	$ git checkout debian -- && git branch -D master

3. Create a fresh "master" branch from the current "debian":

	$ git checkout --orphan master

   This also switches to that new "master".

4. Merge the "upstream" branch into "master":

	$ git merge upstream

5. Build the binary package(s) via

   	$ gbp buildpackage

If it works, the package files are in /tmp/gbp4hpe (see debian/gbp.conf)

For a great reference to the build toolchain, have a look at

https://people.debian.org/~stapelberg/2016/11/25/build-tools.html


debian/changelog
tm-librarian (1.35-6) unstable; urgency=medium

  * Add SDFlex/SDSM configuration validator as standalone command

 -- Rocky Craig <rocky.craig@hpe.com>  Thu, 6 Jun 2019 14:42:42 -0700

tm-librarian (1.35-5) unstable; urgency=medium

  * Tweaks for small FAME FAM (under 1G)
  * Startup scanning improvements
  * Defaults files upgraded

 -- Rocky Craig <rocky.craig@hpe.com>  Thu, 2 May 2019 14:42:42 -0700

tm-librarian (1.35-4) unstable; urgency=medium

  * Allow SD990/Flex to have FAM-less "nodes"
  * Refactor BooksIGInterpretation (BII) enum into a full singleton class
  * Add lslfs script to librarian package

 -- Rocky Craig <rocky.craig@hpe.com>  Tue, 8 Jan 2019 14:42:42 -0700

tm-librarian (1.35-3) unstable; urgency=medium

  * Subsume tm-utils and tm-lmp packages into tm-librarian
  * Clean up some lintian warnings

 -- Rocky Craig <rocky.craig@hpe.com>  Sat, 17 Mar 2018 14:42:42 -0600

tm-librarian (1.35-2) unstable; urgency=medium

  * Uninitialized variable during MFT processing from tm-manifest setup tmconfig

 -- Rocky Craig <rocky.craig@hpe.com>  Tue, 06 Mar 2018 06:47:38 -0700

tm-librarian (1.34-1) unstable; urgency=medium

  * Enable non-default books policies on SD Flex, fix book_register bug on MFT
  * Enable multiple physical address ranges on SD Flex

 -- Greg Pearson <greg.pearson@hpe.com>  Tue, 06 Feb 2018 12:47:38 -0700

tm-librarian (1.32-1) unstable; urgency=medium

  * Fix book_register to work with simple ini file format

 -- Greg Pearson <greg.pearson@hpe.com>  Tue, 14 Nov 2017 14:46:20 -0700

tm-librarian (1.31-1) unstable; urgency=medium

  * Fix database corruption during shelf rename and add new data fields for ED display
  * Add support for 990 systems with single physical address range per partition
  * Fix overlapping message issue

 -- Greg Pearson <greg.pearson@hpe.com>  Wed, 08 Nov 2017 10:35:03 -0700

tm-librarian (1.27-1) unstable; urgency=medium

  * Fix heartbeat unschedule issue
  * Fix file open caching issue
  * Add support for directories

 -- Keith Packard <keithp@keithp.com>  Tue, 15 Aug 2017 09:48:56 -0700

tm-librarian (1.24-1) unstable; urgency=medium

  * Add cpu_percent and rootfs_percent to heartbeat data
  * maptrap updates and read-only access to librarian database for LMP
  * add leading slash to coordinate in LMP allocated routine
  * add fsync to book zeroing, updates to maptrap
  * Handle node death and new maptrap features
  * Fixes to LMP for coordinate syntax changes
  * Added GPL2 headers
  * Create /var/hpetm directory, bug fixes
  * Change /lfs permissions to 777 on creation
  * Create /usr/bin/check-book.py softlink, add services to tmconfig, bug fixes

 -- Keith Packard <keithp@keithp.com>  Thu, 06 Jul 2017 14:30:14 -0700

tm-librarian (1.15-1) unstable; urgency=medium

  * Merge recent librarian changes

 -- Keith Packard <keithp@keithp.com>  Wed, 05 Jul 2017 13:16:04 -0700

tm-librarian (1.14-1) unstable; urgency=medium

  * Set LMP daemon working directory to python tm_librarian dir

 -- Keith Packard <keithp@keithp.com>  Fri, 27 Jan 2017 14:39:02 -0800

tm-librarian (1.13-2) unstable; urgency=low

  * Split into per-daemon packages

 -- Keith Packard <keithp@keithp.com>  Sun, 27 Nov 2016 16:14:46 -0800

tm-librarian (1.13-1) unstable; urgency=low

  * Package for unstable

 -- Keith Packard <packard@hpe.com>  Tue, 15 Nov 2016 14:18:13 -0800


frdnode.py
#!/usr/bin/python3

# Copyright 2017 Hewlett Packard Enterprise Development LP

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2 as
# published by the Free Software Foundation.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License along
# with this program.  If not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

# Full Rack Demo (FRD) for 2016
# Max of one rack of eight enclosures of ten nodes: exclude rack.  FRD
# has 4T per node, or 512 books per node, or 128 books per media controller.

# Chipset ERS 2.02 section 2.21.3.1 option 1: CID== enc[3]:node[4]:subCID[4]
# making an 11-bit CID.  Media controllers are  GenZ responders 1000b - 1111b
# (0x8 - 0xF) but only first 4 # used in FRD (0x8 - 0xB).  "module_size" is
# a phrase from the ERS; there it's in bytes, here it's in books.

# __repr__ functions are really about easier debugging than intended purpose

import os
from pdb import set_trace

from genericobj import GenericObject

#--------------------------------------------------------------------------
# Commonly imported file.  How is the "intlv_group" field in the "books"
# table interpreted (and ultimately, the ID field itself)?  Look at the
# mode in bits 19-16 to interpret the value.  This use of 20 bits means
# the smallest book size is 2^21 or two megabytes, which is also the
# size of an x86 huge page.

class BooksIGInterpretation():
    VALUE_MASK = 0xffff     # Lower 16 bits
    MODE_SHIFT = 16         # Embed the mode above VALUE_MASK
    MODE_MASK = 0xf         # 16 modes max
    MODE_LZA = 0            # Legacy mode, it really is the IG
    MODE_PHYSADDR = 1       # Physaddr mode for 990x and SD Flex
    MODE_INVALID = 0xf      # Okay, 15 modes max
    IG_MASK = 0xff          # 256 IGs max

#--------------------------------------------------------------------------


class FRDnodeID(object):
    """Reverse-calculation of a node ID 1-80 from enc 1-8  and node 1-10."""

    @property
    def node_id(self):
        return (self.rack - 1) * 80 + (self.enc - 1) * 10 + self.node

    @property
    def REN(self):
        '''Rack # Enc # Node # as a string'''
        return 'r%de%dn%d' % (self.rack, self.enc, self.node)

    def __eq__(self, other):
        return self.node_id == other.node_id

    def __hash__(self):
        return self.node_id

#--------------------------------------------------------------------------


class FRDFAModule(FRDnodeID):
    '''One Media Controller (MC) and the books of NVM behind it.'''

    MC_STATUS_OFFLINE = 0
    MC_STATUS_ACTIVE = 1

    def __init__(self, STRorCID=None, enc=None, node=None, ordMC=None,
                 module_size_books=0):
        self.module_size_books = module_size_books
        if STRorCID is not None:
            '''Break apart an encoded string, or just return integer'''
            try:
                enc, node, ordMC = STRorCID.split(':')
                enc = int(enc)
                node = int(node)
                ordMC = int(ordMC)
            except Exception as e:
                # Values in a raw CID are zero based
                enc = (((STRorCID >> 9) & 0x7) + 1)   # 3 bits
                node = (((STRorCID >> 4) & 0xF) + 1)  # 4 bits
                ordMC = STRorCID & 0x3                # 2 LSB in FRD
        else:
            assert enc is not None and node is not None and ordMC is not None

        assert 1 <= enc <= 8, 'Bad enclosure value'
        assert 1 <= node <= 10, 'Bad node value'
        assert 0 <= ordMC <= 3, 'Bad ordMC value'
        self.rack = 1   # FRD, just practicing math elsewhere
        self.enc = enc
        self.node = node
        self.ordMC = ordMC
        self.rawCID = (enc - 1) << 9 | (node - 1) << 4 | (8 + ordMC)
        self.coordinate = 'MemoryBoard/1/MediaController/%d' % (ordMC + 1)

    def __str__(self):
        return 'node_id %2d: %d:%d:%d (%d books)' % (
            self.node_id, self.enc, self.node, self.ordMC,
            self.module_size_books)

    def __repr__(self):
        return self.__str__()

    def __sub__(self, other):
        '''Return the NGMI hop count between two media controllers'''
        if self.enc != other.enc:
            return 5

        # same enclosure
        if self.node != other.node:
            return 3

        # same enclosure, same node
        assert self.ordMC != other.ordMC, 'Psychotic self-talk'
        return 1

#--------------------------------------------------------------------------


class MCCIDlist(object):
    '''Better than a plain list becauase of __repr__ and __str__'''

    def __init__(self, rawCIDlist=None, module_size_books=0):
        if rawCIDlist is None:
            self.mediaControllers = [ ]
            return
        assert len(rawCIDlist) <= 8, 'CID list element count > 8'
        self.mediaControllers = [
            FRDFAModule(STRorCID=c, module_size_books=module_size_books)
                     for c in rawCIDlist
        ]

    def __repr__(self):
        return str([ '0x%x' % cid.rawCID for cid in self.mediaControllers ])

    def __str__(self):
        tmp = [ str(cid) for cid in self.mediaControllers ]
        return str(tmp)

    # Duck-type a list

    def __getitem__(self, index):
        return self.mediaControllers[index]

    def __iter__(self):
        return iter(self.mediaControllers)

    def __len__(self):
        return len(self.mediaControllers)

    def append(self, newFAModule):
        assert len(self.mediaControllers) < 8, 'Max length exceeded'
        self.mediaControllers.append(newFAModule)

#--------------------------------------------------------------------------


class FRDnode(FRDnodeID):

    SOC_STATUS_OFFLINE = 0
    SOC_STATUS_ACTIVE = 1

    # Was 300.0 until we got rid of Monasca, which was driving the
    # librarian of each node at 2 Hz (80 Hz on 40 nodes).  Assuming a
    # very steady state, sampling the cluster once every 5-10 seconds
    # should be good, let's go five.  Reverse Nyquist says sample here
    # at 10, to get an average cluster update of 5.   That's 4 Hz against
    # the librarian from 40 nodes.
    SOC_HEARTBEAT_SECS = 10.0

    def __init__(self, node, enc=None, module_size_books=0, autoMCs=True):
        self.nvm_physaddr = 0   # duck-type to match tmconfig.py (990 FAM)
        self._hostname = None
        node_id = None
        if enc is None:
            try:
                # encoded string
                rack, enc, node = node.split(':')
                rack = int(rack)
                enc = int(enc)
                node = int(node)
            except Exception as e:
                # old school, node is enumerator 1-80
                assert 1 <= node <= 80, 'Bad node enumeration value'
                node_id = node
                n = node - 1   # modulus math
                node = (n % 10) + 1
                enc = ((n % 80) // 10) + 1
                rack = (n // 80) + 1    # always 1
        else:
            rack = 1    # FRD: 1 rack

        assert 1 == rack, 'Bad rack value'
        assert 1 <= enc <= 8, 'Bad enclosure value'
        assert 1 <= node <= 10, 'Bad node value'
        self.node = node
        self.enc = enc
        self.rack = rack
        if node_id is not None:     # property check
            assert self.node_id == node_id

        # Duck-type spoof the objects generated from a JSON TMCF.  A better
        # long-term approach is to meld this file's objects into tmconfig.py.
        # The value is for QEMU (52:54), FAME (42), to trigger node ID
        # (matching three octects).

        self.coordinate = 'Node/%d' % (self.node)
        self.serialNumber = self.physloc
        self.soc = GenericObject(
            macAddress='52:54:42:%02d:%02d:%02d' % (
                self.node_id, self.node_id, self.node_id),
            coordinate='SocBoard/1/Soc/1',
            tlsPublicCertificate='NotToday'
        )

        # Media controllers. If not auto-generated (see the chipset ERS)
        # there's probably a custom module_size_books in the INI file.
        if not autoMCs:
            self.mediaControllers = []
            return

        # Could also loop on (enc=, node=, ordMC=) constructor but this
        # is how it will appear in INI file for [InterleaveGroups] section.
        self.mediaControllers = MCCIDlist(
            [ '%d:%d:%d' % (enc, node, ordMC) for ordMC in range(4) ],
            module_size_books
        )

    # FIXME: TMConfig should use this class
    @property
    def hostname(self):
        if self._hostname is None:
            # MFT/FRD: rack is always "1", or words like "A1.above_floor"
            self._hostname = 'node%02d' % self.node_id
        return self._hostname

    @hostname.setter
    def hostname(self, value):
        self._hostname = str(value)

    @property
    def physloc(self):
        return '%(rack)d:%(enc)d:%(node)d' % self.__dict__

    def __str__(self):
        return 'node_id %2d: %-6s %s' % (self.node_id, self.physloc,
        self.mediaControllers)

    def __repr__(self):
        return self.__str__()

#--------------------------------------------------------------------------
# Interleave groups, using the abbreviation from the chipset ERS.


class FRDintlv_group(object):

    def __init__(self, groupId, mediaControllers):
        assert 0 <= groupId < 128, 'intlv_group ID out of range 0-127'
        self.groupId = groupId
        self.mediaControllers = mediaControllers
        # Real machine hardware only has 13 bits of book number in an LZA
        assert self.total_books <= 8192, 'book count too large'

    def __str__(self):
        return '%-3s %s' % (self.groupId, self.mediaControllers)

    def __repr__(self):
        return self.__str__()

    @property
    def total_books(self):
        '''Supersedes ig_gap calculations for flat-space emulations'''
        total_books = 0
        for mc in self.mediaControllers:
            total_books += mc.module_size_books
        return total_books
